# Pembersihan Outlier

In [ ]:
import pandas as pd
from sqlalchemy import create_engine

MYSQL_HOST = "database-aplikasi-saya-mysql-muhammadazhariii1209-4402.j.aivencloud.com"
MYSQL_PORT = 14467
MYSQL_USER = "avnadmin"
MYSQL_PASSWORD = "AVNS_-m75fnNFJLYQ3JbqMc4"
MYSQL_DATABASE = "defaultdb"

PG_HOST = "database-aplikasi-saya-muhammadazhariii1209-4402.j.aivencloud.com"
PG_PORT = 14467
PG_USER = "avnadmin"
PG_PASSWORD = "AVNS_cAY9-ITh6Sy3eujURVI"
PG_DATABASE = "defaultdb"

mysql_engine = create_engine(
    f"mysql+pymysql://{MYSQL_USER}:{MYSQL_PASSWORD}@{MYSQL_HOST}:{MYSQL_PORT}/{MYSQL_DATABASE}"
)

pg_engine = create_engine(
    f"postgresql+psycopg2://{PG_USER}:{PG_PASSWORD}@{PG_HOST}:{PG_PORT}/{PG_DATABASE}"
)

mysql_query = "SELECT * FROM iris_mysql;"
df_mysql = pd.read_sql(mysql_query, mysql_engine)

pg_query = "SELECT * FROM iris_postgre;"
df_postgres = pd.read_sql(pg_query, pg_engine)

print("Data dari MySQL:")
print(df_mysql.head())

print("\nData dari PostgreSQL:")
print(df_postgres.head())

In [ ]:
import os
mysql_query = "SELECT * FROM iris_mysql;"
df_mysql = pd.read_sql(mysql_query, mysql_engine)

pg_query = "SELECT * FROM iris_postgre;"
df_postgres = pd.read_sql(pg_query, pg_engine)

# Print hasil query
print("Data dari MySQL:")
print(df_mysql.head())

print("\n Data dari PostgreSQL:")
print(df_postgres.head())

#Menghapus file yang di upload sebelumnya
if os.path.exists(".env"):
    os.remove(".env")


df_combined = pd.merge(df_postgres, df_mysql.drop(columns=['Class']), on="id", how="inner")

print("\nData gabungan:")
print(df_combined)

## Seleksi outlier menggunakan metode ABOD

In [3]:
from pycaret.datasets import get_data
from pycaret.anomaly import *

# df_combined = dataset 

# Use only the feature columns for anomaly detection
anomaly = df_combined[['sepal length', 'sepal width', 'petal length', 'petal width']]

exp_name = setup(data=anomaly)
abod = create_model('abod', fraction=0.05)
abod_df = assign_model(abod) 

outlier_df = abod_df[abod_df['Anomaly'] == 0]
outlier_df

sepal length  sepal width  petal length  petal width  Anomaly  \
0             5.1          3.5           1.4          0.2        0   
1             4.9          3.0           1.4          0.2        0   
2             4.7          3.2           1.3          0.2        0   
3             4.6          3.1           1.5          0.2        0   
4             5.0          3.6           1.4          0.2        0   
..            ...          ...           ...          ...      ...   
145           6.7          3.0           5.2          2.3        0   
146           6.3          2.5           5.0          1.9        0   
147           6.5          3.0           5.2          2.0        0   
148           6.2          3.4           5.4          2.3        0   
149           5.9          3.0           5.1          1.8        0   

     Anomaly_Score  
0      -556.251421  
1      -400.000928  
2       -93.421993  
3       -99.229221  
4       -82.176201  
..             ...  
145     -13.831461  
146      -9.110201  
147     -20.571707  
148      -6.345400  
149     -28.205616  

[142 rows x 6 columns]

* seharusnya pada data diatas hanya tersisa 142 data karean code diatas mendeteksi anomaly=0

In [4]:
# Simpan hasil ke file CSV
outlier_df.to_csv("outlier_abod.csv", index=False)

Tapi pada csv diatas sudah bersih 142 data

## Seleksi outlier menggunakan metode KNN 

In [5]:
from pycaret.datasets import get_data
from pycaret.anomaly import *

# Use only the feature columns for anomaly detection
anomaly = df_combined[['sepal length', 'sepal width', 'petal length', 'petal width']]

exp_name = setup(data=anomaly)
knn = create_model('knn', fraction=0.05)
knn_df = assign_model(knn) 

knn_outlier_df = knn_df[abod_df['Anomaly'] == 0]
knn_outlier_df

sepal length  sepal width  petal length  petal width  Anomaly  \
0             5.1          3.5           1.4          0.2        0   
1             4.9          3.0           1.4          0.2        0   
2             4.7          3.2           1.3          0.2        0   
3             4.6          3.1           1.5          0.2        0   
4             5.0          3.6           1.4          0.2        0   
..            ...          ...           ...          ...      ...   
145           6.7          3.0           5.2          2.3        0   
146           6.3          2.5           5.0          1.9        0   
147           6.5          3.0           5.2          2.0        0   
148           6.2          3.4           5.4          2.3        0   
149           5.9          3.0           5.1          1.8        0   

     Anomaly_Score  
0         0.141421  
1         0.173205  
2         0.264575  
3         0.264575  
4         0.244949  
..             ...  
145       0.374166  
146       0.479583  
147       0.387298  
148       0.624500  
149       0.360555  

[142 rows x 6 columns]

In [6]:
# Simpan hasil ke file CSV
knn_outlier_df.to_csv("outlier_knn.csv", index=False)

## Seleksi outlier menggunakan metode LOF

In [7]:
from pycaret.datasets import get_data
from pycaret.anomaly import *

# Use only the feature columns for anomaly detection
anomaly = df_combined[['sepal length', 'sepal width', 'petal length', 'petal width']]

exp_name = setup(data=anomaly)
lof = create_model('lof', fraction=0.05)
lof_df = assign_model(lof) 

lof_outlier_df = lof_df[abod_df['Anomaly'] == 0]
lof_outlier_df

sepal length  sepal width  petal length  petal width  Anomaly  \
0             5.1          3.5           1.4          0.2        0   
1             4.9          3.0           1.4          0.2        0   
2             4.7          3.2           1.3          0.2        0   
3             4.6          3.1           1.5          0.2        0   
4             5.0          3.6           1.4          0.2        0   
..            ...          ...           ...          ...      ...   
145           6.7          3.0           5.2          2.3        0   
146           6.3          2.5           5.0          1.9        0   
147           6.5          3.0           5.2          2.0        0   
148           6.2          3.4           5.4          2.3        0   
149           5.9          3.0           5.1          1.8        0   

     Anomaly_Score  
0         0.976302  
1         1.008758  
2         1.019841  
3         1.049882  
4         0.958473  
..             ...  
145       0.978474  
146       1.004232  
147       0.980847  
148       1.021819  
149       1.011326  

[142 rows x 6 columns]

In [8]:
# Simpan hasil ke file CSV
lof_outlier_df.to_csv("outlier_lof.csv", index=False)